[CIVIS-9525] correlate ci cd docs (#22303)
DataDog/documentation: content/en/continuous_delivery/deployments/argocd.md

    dd_customtags: "region:us1-east, team:backend"
```

## Correlate deployments with CI pipelines

If Argo CD deployments are triggered from a CI pipeline, you may correlate the deployment execution and the pipeline,

> allowing to check the associated deployments to a pipeline and the pipeline that triggered a deployment from Datadog.

Reviewer: Suggesting to break this sentence up into two for ease of reading. Additionally, could you provide some more context on the second sentence? For example, what value would a user get from correlating the deployments to a pipeline?

> allowing to check the associated deployments to a pipeline and the pipeline that triggered a deployment from Datadog.

Author: Good point, when the frontend is finished I will attach an screenshot here

> In this moment, not all the setups are supported. To be able to make the correlation your setup has to be similar to this:

Reviewer: ```suggestion
To correlate your deployments with your pipeline, your CI pipeline must make its changes to the Git repository with `git commit` as shown below:
```

Reviewer: As a general rule, the documentation emphasizes what **is** supported, rather than what is **not** supported.

Since not all users are able to view images, I added a suggestion here to describe the key requirement for this setup to work. Feel free to make changes if my suggestion is not accurate.

{{< img src="ci/cd-argocd-ci-correlation-setup-git.png" alt="Triggering Argo CD deployments using git" style="width:100%;">}}

> If you are using [argo cd image updater][14], this command does not work since it relies on making the changes on
> the configuration repository using `git commit`.

Reviewer: ```suggestion
**Note**: [Argo CD Image Updater][14] does not make changes to
the configuration repository with `git commit`, and as such is incompatible with this setup.
```

> If you are using a similar setup as the one described above, then the correlation can be done by calling
> `datadog-ci deployment correlate` ([command syntax][15]) just before pushing the changes to the configuration repository:

Reviewer: ```suggestion
If your setup is similar to the one described above, run the `datadog-ci deployment correlate` command before pushing the changes to the configuration repository. See the [command syntax][15] for additional details:
```

```yaml
- job: JobToUpdateConfigurationRepository
  run: |
    # Update the configuration files
    ...
    git commit
    # Correlate the deployment with the CI pipeline
    datadog-ci deployment correlate --provider argocd
    git push
```

> Note: the code snippet above is just an example. You should adapt it to your CI provider syntax.

Reviewer: ```suggestion
**Note**: the code snippet above is an example. Adapt it to your CI provider syntax as necessary.
```

By default, the git metadata reported in deployments will be related to the repository that Argo CD monitors. However, a common setup is to separate the source code repository from the repository Argo CD monitors (the configuration repository), as the [Argo CD best practices states][17]. Then, to update the configuration repository when changes in the source repository happen, it is common to use automated commits from a CI pipeline. The following diagram represents an example of this kind of setup:

{{< img src="ci/cd-argocd-ci-correlation-setup-git.png" alt="Triggering Argo CD deployments using git" style="width:100%;">}}

To associate source code git information to the Argo CD deployments, it is required to run the `datadog-ci deployment correlate` command before pushing the changes to the configuration repository. See the [command syntax][14] for additional details:

```yaml
- job: JobToUpdateConfigurationRepository
  run: |
    # Update the configuration files
    ...
    git commit
    # Correlate the deployment with the CI pipeline
    datadog-ci deployment correlate --provider argocd
    git push
```

> Even if the same repository serves as the source code repository and configuration repository, running this command is still required since it also gathers the information required to perform the correlation between the pipeline and the deployment.

Reviewer: Suggesting to make this a **Note** for visibility

> Once this command runs in the CI, the deployment contains git metadata from the application repository instead of the configuration repository as well as enough information to show the associated pipeline. See the section ["visualize correlated pipelines and deployments"][18] to get more details about the visualization.

Reviewer: I'm open to feedback on this if you'd rather not add a **Validation** subsection, or if you have any suggested improvements for the wording.

My reasons for adding the **Validation** section are:
- It visually breaks up the content and clarifies what it's about
- It's used in other parts of the documentation

Thanks in advance for any feedback!

> Once this command runs in the CI, the deployment contains git metadata from the application repository instead of the configuration repository as well as enough information to show the associated pipeline. See the section ["visualize correlated pipelines and deployments"][18] to get more details about the visualization.

Author: Thanks! Happy to add the validation heading there, I think it is a good idea to break up the content since it makes it more readable

    dd_k8s_cluster: example-cluster
```

## Visualize deployments in Datadog

The [**Deployments**][6] and [**Executions**][7] pages populate with data after a deployment is executed. For more information, see [Search and Manage][9] and [CD Visibility Explorer][10].

## Correlate deployments with CI pipelines

If Argo CD deployments are triggered from a CI pipeline, you may correlate the deployment execution and the pipeline,
allowing to check the associated deployments to a pipeline and the pipeline that triggered a deployment from Datadog. See the [section below](#visualize-correlated-pipelines-and-deployments) to see how the visualization looks in Datadog.

> In this moment, not all the setups are supported. To be able to make the correlation your setup needs to make changes to the repository Argo CD is monitoring using `git commit` in your CI. The following diagram represents an example of this kind of setup:

Reviewer 2: Based on my comments above, I'd mention the use case we support at the beginning (app repo, config repo, automated commit) and remove this sentence

## Correlate deployments with CI pipelines

By default, the Git metadata reported in deployment events is associated with the repository that Argo CD monitors. However, a common setup is to:
- Have an application repository, storing the source code, and a configuration repository, storing the Kubernetes manifests. Then, configure Argo CD to monitor the configuration repository, as outlined in the [Argo CD Best Practices page][17].
- When a change occurs in the application repository, perform an automated commit that updates the configuration repository (for example, changing the current image of a Kubernetes resource).

The following diagram represents an example of this kind of setup:

{{< img src="ci/cd-argocd-ci-correlation-setup-git.png" alt="Triggering Argo CD deployments using git" style="width:100%;">}}

> In this case, you can replace the Git metadata reported in the deployment with the one of the application repository instead of the configuration repository. This allows you to connect the deployments performed by Argo CD and the related CI pipeline runs on the application repository.

Reviewer: ```suggestion
In this case, you can replace the Git metadata reported in the deployment with the metadata of the application repository instead of the configuration repository. This allows you to connect the deployments performed by Argo CD and the related CI pipeline runs on the application repository.
```

    git push
```

> **Note**: Even in case a single repository is used to store both the source code and the Kubernetes manifest, running this command is still required to correctly associate deployments and CI pipelines.

Reviewer: ```suggestion
**Note**: Even if a single repository is used to store both the source code and the Kubernetes manifest, running this command is still required to correctly associate deployments and CI pipelines.
```

### Validation

> If the command has been correctly run, deployments contain Git metadata from the application repository instead of the configuration repository. Also, the deployment executions view will contain a new **Pipeline** tab representing the related CI pipeline trace.

Reviewer: ```suggestion
If the command has been correctly run, deployments contain Git metadata from the application repository instead of the configuration repository. Also, the deployment executions view now contains a new **Pipeline** tab representing the related CI pipeline trace.
```
